src/agora/txrx/txrx_argos.cc

```diff
@@ -89,69 +118,79 @@ struct Packet* PacketTXRX::RecvEnqueueArgos(int tid, int radio_id,
     }
   }
-  return pkt[0];
+  return pkt;
 }
 
 int PacketTXRX::DequeueSendArgos(int tid) {
-  auto& c = cfg_;
-  EventData event;
-  if (task_queue_->try_dequeue_from_producer(*tx_ptoks_[tid], event) == false) {
+  std::array<EventData, 4> event;
+  if (task_queue_->try_dequeue_bulk_from_producer(*tx_ptoks_[tid], event.data(),
+                                                  cfg_->NumChannels()) == 0) {
     return -1;
   }
 
   // std::printf("tx queue length: %d\n", task_queue_->size_approx());
   assert(event.event_type_ == EventType::kPacketTX);
 
-  size_t ant_id = gen_tag_t(event.tags_[0]).ant_id_;
-  size_t frame_id = gen_tag_t(event.tags_[0]).frame_id_;
-  size_t symbol_id = gen_tag_t(event.tags_[0]).symbol_id_;
+  size_t frame_id = gen_tag_t(event.at(0).tags_[0]).frame_id_;
+  size_t symbol_id = gen_tag_t(event.at(0).tags_[0]).symbol_id_;
+  size_t ant_id = gen_tag_t(event.at(0).tags_[0]).ant_id_;
+  size_t radio_id = ant_id / cfg_->NumChannels();
 
   size_t dl_symbol_idx = cfg_->Frame().GetDLSymbolIdx(symbol_id);
-  size_t offset =
-      (c->GetTotalDataSymbolIdxDl(frame_id, dl_symbol_idx) * c->BsAntNum()) +
-      ant_id;
-
-  // TODO: support nChannels > 1
-  void* txbuf[2];
-  int n_channels = c->NumChannels();
-  int ch = ant_id % n_channels;
-
-  if ((c->Frame().IsRecCalEnabled() == true) &&
-      (symbol_id == c->Frame().GetDLSymbol(0)) && (ant_id != c->RefAnt())) {
-    std::vector<std::complex<int16_t>> zeros(c->SampsPerSymbol(),
+  size_t offset = (cfg_->GetTotalDataSymbolIdxDl(frame_id, dl_symbol_idx) *
+                   cfg_->BsAntNum()) +
+                  ant_id;
+
+  // Transmit downlink calibration (array to ref) pilot
+  void* caltxbuf[2];
+  if ((cfg_->Frame().IsRecCalEnabled() == true) &&
+      (symbol_id == cfg_->Frame().GetDLSymbol(0)) &&
+      (radio_id != cfg_->RefRadio())) {
+    std::vector<std::complex<int16_t>> zeros(cfg_->SampsPerSymbol(),
                                              std::complex<int16_t>(0, 0));
-    for (size_t s = 0; s < c->AntPerGroup(); s++) {
-      txbuf[ch] = (frame_id % c->AntGroupNum() == ant_id / c->AntPerGroup() &&
-                   s == ant_id % c->AntPerGroup())
-                      ? c->PilotCi16().data()
-                      : zeros.data();
-      long long frame_time = ((long long)(frame_id + TX_FRAME_DELTA) << 32) |
-                             (c->Frame().GetDLCalSymbol(s) << 16);
-      radioconfig_->RadioTx(ant_id / n_channels, txbuf, 1, frame_time);
+    for (size_t s = 0; s < cfg_->RadioPerGroup(); s++) {
+      bool calib_turn = (frame_id % cfg_->RadioGroupNum() ==
+                             radio_id / cfg_->RadioPerGroup() &&
+                         s == radio_id % cfg_->RadioPerGroup());
+      for (size_t ch = 0; ch < cfg_->NumChannels(); ch++) {
+        caltxbuf[ch] = calib_turn ? cfg_->PilotCi16().data() : zeros.data();
+        if (cfg_->NumChannels() > 1) caltxbuf[1 - ch] = zeros.data();
```

Comment:
If cfg_->NumChannels() > 1  and NumChannels != 2 will cause a memory problem.  Should probably add a check somewhere.